### PR TITLE
Issue 4528 - Fix cn=monitor SCOPE_ONE search

### LINKED ISSUE #4528: cn=monitor is throwing err=32 with scope: -s one
**Issue Description**
While doing a ldapsearch on "cn=monitor" is throwing err=32 with -s one.

**Package Version and Platform:**
 - Package and version: 389-ds-base master branch

**Steps to Reproduce**
1. ldapsearch -LLL -Y EXTERNAL -H 'ldapi://%2fvar%2frun%2fslapd-test.socket' -b "cn=monitor" -s one
No such object (32)

**Expected results**
Content is found.

### PR DESCRIPTION
**Pull request by droideck**
Bug Description: While doing a ldapsearch on "cn=monitor" is throwing
err=32 with -s one.

Fix Description: 'cn=monitor' is not a real entry so we should not
trying to check if the searched suffix (cm=monitor or its children)
belongs to the searched backend.

Fixes: #4528

Reviewed by: ?

### CODE REVIEW

File: ldap/servers/slapd/opshared.c [hunk 1]
```diff
@@ -809,6 +812,7 @@ op_shared_search(Slapi_PBlock *pb, int send_result)
                     }
                 }
             }
+            slapi_sdn_free(&monitorsdn);
```
Comment (mreynolds389) [inline ldap/servers/slapd/opshared.c]: You also need to free this before we "goto next_be" at line 794

**Comment by mreynolds389** [comment]
You can also avoid some allocations by using a local variable for monitorsdn:

Slapi_DN monitorsdn = {0};
slapi_sdn_init_dn_byval(&monitorsdn, "cn=monitor");
/* do suffix check */
slapi_sdn_done(&monitorsdn);

**Comment by droideck** [comment]
Okay, posting the working fix but...
I am really not sure why but this approach has failed for me:

```
Slapi_DN monitorsdn = {0};
slapi_sdn_init_dn_byval(&monitorsdn, "cn=monitor");
/* do suffix check */
slapi_sdn_done(&monitorsdn);
```

The way it failed - it was not starting after installation- completely silently! No error logs, not journalctl info...

This is the patch I've used:

```
From ccc477f024717055e2430b83f3266579ab64418a Mon Sep 17 00:00:00 2001
From: Simon Pichugin <spichugi@redhat.com>
Date: Tue, 12 Jan 2021 17:03:26 +0100
Subject: [PATCH] Apply Mark's comments

---
 ldap/servers/slapd/opshared.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/ldap/servers/slapd/opshared.c b/ldap/servers/slapd/opshared.c
index b8a25955b..6876ced25 100644
--- a/ldap/servers/slapd/opshared.c
+++ b/ldap/servers/slapd/opshared.c
@@ -245,6 +245,7 @@ op_shared_search(Slapi_PBlock *pb, int send_result)
     Slapi_Entry *referral = NULL;
     char *proxydn = NULL;
     char *proxystr = NULL;
+    Slapi_DN *monitorsdn = {0};
     int proxy_err = LDAP_SUCCESS;
     char *errtext = NULL;
     int nentries, pnentries;
@@ -767,7 +768,7 @@ op_shared_search(Slapi_PBlock *pb, int send_result)
              * -> don't change the search parameters in pblock
              * Also, we skip this block for 'cn=monitor' search and its subsearches
              * as they are done by callbacks from monitor.c */
-            Slapi_DN *monitorsdn = slapi_sdn_new_ndn_byval("cn=monitor");
+            slapi_sdn_init_dn_byval(&monitorsdn, "cn=monitor");
             if ((be_suffix != NULL) && !slapi_sdn_issuffix(basesdn, monitorsdn)) {
                 if ((be_name == NULL) && (scope == LDAP_SCOPE_ONELEVEL)) {
                     /* one level searches
@@ -812,7 +813,6 @@ op_shared_search(Slapi_PBlock *pb, int send_result)
                     }
                 }
             }
-            slapi_sdn_free(&monitorsdn);

             slapi_pblock_set(pb, SLAPI_BACKEND, be);
             slapi_pblock_set(pb, SLAPI_PLUGIN, be->be_database);
@@ -1057,6 +1057,7 @@ free_and_return_nolock:

     slapi_ch_free_string(&proxydn);
     slapi_ch_free_string(&proxystr);
+    slapi_sdn_done(&monitorsdn);

 #ifdef SYSTEMTAP
     STAP_PROBE(ns-slapd, op_shared_search__return);
--
2.29.2
```

File: ldap/servers/slapd/opshared.c [hunk 2]
```diff
@@ -764,8 +765,10 @@ op_shared_search(Slapi_PBlock *pb, int send_result)
             }
         } else {
             /* be_suffix null means that we are searching the default backend
-             * -> don't change the search parameters in pblock */
-           if (be_suffix != NULL) {
+             * -> don't change the search parameters in pblock
+             * Also, we skip this block for 'cn=monitor' search and its subsearches
+             * as they are done by callbacks from monitor.c */
+            if ((be_suffix != NULL) && !slapi_sdn_issuffix(basesdn, monitorsdn)) {
```
Comment (Firstyear) [inline ldap/servers/slapd/opshared.c]: This logic may be clearer as:

```
if (!( be_suffix == NULL) || slapi_sdn_issuffix() ))
```

**Review by Firstyear — CHANGES_REQUESTED** [review]
Does slapi_sdn_done try to free the pointer? because this is on the stack not the heap it may be trying to free it incorrectly, which could cause a crash/failure.

**Comment by mreynolds389** [comment]
Sorry @droideck that is not correct and probably leaks memory.  Let me put something together for you in a little bit.  Stay tuned...

**Comment by mreynolds389** [comment]
@droideck this patch works for me:

```
diff --git a/ldap/servers/slapd/opshared.c b/ldap/servers/slapd/opshared.c
index c8b45bfdb..114ca4863 100644
--- a/ldap/servers/slapd/opshared.c
+++ b/ldap/servers/slapd/opshared.c
@@ -240,6 +240,7 @@ op_shared_search(Slapi_PBlock *pb, int send_result)
     int rc = 0;
     int internal_op;
     Slapi_DN *basesdn = NULL;
+    Slapi_DN monitorsdn = {0};
     Slapi_DN *sdn = NULL;
     Slapi_Operation *operation = NULL;
     Slapi_Entry *referral = NULL;
@@ -764,8 +765,11 @@ op_shared_search(Slapi_PBlock *pb, int send_result)
             }
         } else {
             /* be_suffix null means that we are searching the default backend
-             * -> don't change the search parameters in pblock */
-           if (be_suffix != NULL) {
+             * -> don't change the search parameters in pblock
+             * Also, we skip this block for 'cn=monitor' search and its subsearches
+             * as they are done by callbacks from monitor.c */
+            slapi_sdn_init_dn_byref(&monitorsdn, "cn=monitor");
+            if ((be_suffix != NULL) && !slapi_sdn_issuffix(basesdn, &monitorsdn)) {
                 if ((be_name == NULL) && (scope == LDAP_SCOPE_ONELEVEL)) {
                     /* one level searches
                      * - depending on the suffix of the backend we might have to
@@ -787,8 +791,10 @@ op_shared_search(Slapi_PBlock *pb, int send_result)
                     } else if (slapi_sdn_issuffix(basesdn, be_suffix)) {
                         int tmp_scope = LDAP_SCOPE_ONELEVEL;
                         slapi_pblock_set(pb, SLAPI_SEARCH_SCOPE, &tmp_scope);
-                    } else
+                    } else {
+                        slapi_sdn_done(&monitorsdn);
                         goto next_be;
+                    }
                 }
 
                 /* subtree searches :
@@ -809,7 +815,7 @@ op_shared_search(Slapi_PBlock *pb, int send_result)
                     }
                 }
             }
-
+            slapi_sdn_done(&monitorsdn);
             slapi_pblock_set(pb, SLAPI_BACKEND, be);
             slapi_pblock_set(pb, SLAPI_PLUGIN, be->be_database);
             slapi_pblock_set(pb, SLAPI_SEARCH_RESULT_SET, NULL);
```

**Comment by tbordaz** [comment]
The fix looks valid. 
'cn=monitor' as a dse special suffix is currently handled via fedse registered callbacks (https://github.com/389ds/389-ds-base/blob/master/ldap/servers/slapd/fedse.c#L2921).
Another option would be to add a callback to process scope one. It would leave opshared without 'cn=monitor' specific.

**Comment by droideck** [comment]
> Another option would be to add a callback to process scope one. It would leave opshared without 'cn=monitor' specific.

I think it wouldn't help as 'NO_SUCH_ENTRY' error happens even earlier because it tries to call these functions
`slapi_sdn_isparent(basesdn, be_suffix), slapi_sdn_issuffix(basesdn, be_suffix)` and these calls are not valid for 'cn=monitor' case.
So my fix is about skipping the calls for 'cn=monitor'. And then it successfully completes all of the monitor magic :)

> @droideck this patch works for me:

Okay, I've tried the solution you propose but it was failing (no error, no warning, the instance just silently fails to start) 
But I was using `slapi_sdn_init_dn_byval` instead of `slapi_sdn_init_dn_byref`...
Looking at the functions I am not sure how exactly it can cause a failure so any insight is welcome:)

But anyway, the `_byref` variant works like a charm so please, review. :)

**Comment by mreynolds389** [comment]
@droideck - please do a ASAN run just to make sure nothing is wrong.  If it passes then ACK!

File: ldap/servers/slapd/opshared.c (hunk 2 above)
Comment (droideck) [reply ldap/servers/slapd/opshared.c]: Probably just a typo, but I think you meant  (!((be_suffix == NULL) || slapi_sdn_issuffix())) - the third parenthesis. 
But yeah, I agree, it sounds more cleaner:) Fixed

**Comment by droideck** [comment]
> @droideck - please do a ASAN run just to make sure nothing is wrong. If it passes then ACK!

Looks like there are no leaks. Merging...

**Pull request merged**